### PR TITLE
Adds a callback mechanism for observing mutation of relations

### PR DESCRIPTION
**Pull request by mbarnett**
The concrete use-case for this is allowing ActiveFedora to be aware of when properties become dirty in more cases than it previously could (specifically, this allows it to detect when values are <<'d onto a relation and persist the changes properly, addressing a long-standing bug/PITA). A PR has been opened ActiveFedora to take advantage of this.

(I'm just kind of guessing that develop is the correct branch to PR against?)

### CODE REVIEW

File: lib/active_triples/rdf_source.rb [hunk 1]
```diff
@@ -93,6 +93,8 @@ def initialize(*args, &block)
       resource_uri = args.shift unless args.first.is_a?(Hash)
       @rdf_subject = get_uri(resource_uri) if resource_uri
 
+      callback = args.first.delete(:callback) if args.first.is_a?(Hash)
```
Comment (jcoyne) [inline lib/active_triples/rdf_source.rb]: Can you add YARD documentation about this option. What sort of object is callback expected to be? Could it be renamed to `property_change_callback` or `on_property_change` to be more clear about it's purpose?

File: lib/active_triples/rdf_source.rb [hunk 2]
```diff
@@ -107,6 +109,9 @@ def initialize(*args, &block)
       Array.wrap(self.class.type).each do |type|
         get_values(:type) << type unless get_values(:type).include?(type)
       end
+
+      # This needs to be set last to prevent infinite loops in construction
```
Comment (jcoyne) [inline lib/active_triples/rdf_source.rb]: Can you explain more about this?  How does the infinite loop occur?

File: spec/active_triples/rdf_source_spec.rb [hunk 3]
```diff
@@ -720,4 +720,22 @@ class MyDataModel < ActiveTriples::Schema
       expect { dummy_source.new.test_title }.not_to raise_error
     end
   end
+
+  describe 'callback' do
+    let(:callback) {double}
+
+    before do
+      class TestRdfSource
```
Comment (jcoyne) [inline spec/active_triples/rdf_source_spec.rb]: Can you remove this class in an `after` block, or create an anonymous class (e.g. `Class.new do ... end`)

File: spec/active_triples/rdf_source_spec.rb [hunk 4]
```diff
@@ -720,4 +720,22 @@ class MyDataModel < ActiveTriples::Schema
...
+        include ActiveTriples::RDFSource
+        property :test_title, predicate: RDF::Vocab::DC.title
+      end
+    end
+
+    it 'should be invoked when a property changes' do
```
Comment (jcoyne) [inline spec/active_triples/rdf_source_spec.rb]: use `is` instead of `should be`

**Comment by jcoyne** [comment]
Could this be done by just overriding `<<` in `ActiveFedora::Dirty` without requiring any changes to ActiveTriples?  Then it would work on objects that were dumped/loaded too.

**Comment by mbarnett** [comment]
That was my first thought, but no. Given some object `a`, an instance of class deriving from ActiveFedora::Base with the property title, then `a.title.class` is `ActiveTriples::Relation` and `a.title << ...` invokes `ActiveTriples::Relation#<<`. Only `a << ... ` invokes the `<<` method defined in `ActiveFedora::AttributeMethods::Dirty`.

File: lib/active_triples/rdf_source.rb (hunk 2 above)
Comment (mbarnett) [inline lib/active_triples/rdf_source.rb]: If the callback is in place before line 107, then `get_values(:type) << ...` invokes `set`, which invokes `parent.property_will_change`, which eventually reaches `ActiveModel#attribute_will_change!` which, several calls later, leads to an invocation of https://github.com/mbarnett/active_fedora/blob/f4adcb9f3be4eaba20e489dd4f0010beb8a718cb/lib/active_fedora/fedora_attributes.rb#L46, which hasn't finished constructing, so `@relation` is as yet unset, so it calls the constructor again, and we do this until stack overflow.

I can flesh out that comment to better explain the situation. In general there's a chasing-your-own-tail trap in being notified of mutation in objects that haven't finished constructing.

**Comment by mbarnett** [comment]
I've pushed some changes to address your feedback. Is it worth rebasing to keep this down to a single commit?

**Comment by no-reply** [comment]
Closing this as superseded by #263. If that PR doesn't get merged, I'm happy to reopen.

**Pull request closed without merging**